src/mame/namco/ns10crypt.cpp

```diff
@@ -257,8 +273,106 @@ void ns10_type2_decrypter_device::init(int iv)
+}
+
+void ns10_type2_decrypter_device::device_reset()
+{
+	ns10_decrypter_device::device_reset();
+
+	m_mask = 0;
+	m_previous_cipherwords = 0;
+	m_previous_plainwords = 0;
+}
+
+// type-2 decrypter with a table for the non-linear bits
+const int ns10_type2_decrypter_nonlinear_device::INIT_SBOX[16]{0, 12, 13, 6, 2, 4, 9, 8, 11, 1, 7, 15, 10, 5, 14, 3};
+
+ns10_type2_decrypter_nonlinear_device::ns10_type2_decrypter_nonlinear_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: ns10_decrypter_device(mconfig, NS10_TYPE2_DECRYPTER_NONLINEAR, tag, owner, clock)
+{
+}
+
+ns10_type2_decrypter_nonlinear_device::ns10_type2_decrypter_nonlinear_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock, ns10_crypto_logic &&logic)
+	: ns10_decrypter_device(mconfig, NS10_TYPE2_DECRYPTER_NONLINEAR, tag, owner, clock)
+	, m_logic(std::move(logic))
+{
+	m_logic_initialized = true;
+}
+
+uint16_t ns10_type2_decrypter_nonlinear_device::decrypt(uint16_t cipherword)
+{
+	uint16_t const plainword = cipherword ^ m_mask;
+
+	m_previous_cipherwords <<= 16;
+	m_previous_cipherwords ^= cipherword;
+	m_previous_plainwords <<= 16;
+	m_previous_plainwords ^= plainword;
+
+	m_mask = 0;
+	for (int j = 15; j >= 0; --j)
+	{
+		m_mask <<= 1;
+		m_mask ^= gf2_reduce(m_logic.eMask[j] & m_previous_cipherwords);
+		m_mask ^= gf2_reduce(m_logic.dMask[j] & m_previous_plainwords);
+	}
+	m_mask ^= m_logic.xMask;
+
+	uint8_t nonlinear_bit = BIT(m_nonlinear_region_base[m_nonlinear_count / 8], 7 - (m_nonlinear_count % 8));
+	m_nonlinear_count++;
+	if (m_nonlinear_count >= m_nonlinear_region_size * 8)
+		m_nonlinear_count = 0;
+	m_mask ^= m_logic.nonlinear_calculation(nonlinear_bit);
+
+	return plainword;
+}
+
+void ns10_type2_decrypter_nonlinear_device::init(int iv)
+{
+	if (m_logic.iv_calculation)
+		m_previous_cipherwords = m_logic.iv_calculation(iv);
+	else
+		m_previous_cipherwords = bitswap(INIT_SBOX[iv], 3, 16, 16, 2, 1, 16, 16, 0, 16, 16, 16, 16, 16, 16, 16, 16);
+	m_previous_plainwords = 0;
+	m_mask = 0;
+}
+
+void ns10_type2_decrypter_nonlinear_device::device_start()
+{
+	ns10_decrypter_device::device_start();
+
+	// If the logic isn't initialized then this will just fail, this is a programmer error
+	if (!m_logic_initialized)
+		fatalerror("ns10_type2_decrypter_nonlinear_device: Required logic data for decrypter device not initialized");
+
+	// If the entire purpose for this separate device doesn't exist then just fail, this is a programmer error
+	auto nonlinear_table = memregion("nonlinear_table");
+	if (nonlinear_table == nullptr)
+		fatalerror("ns10_type2_decrypter_nonlinear_device: Non-linear bit lookup table memory region not found");
+
+	save_item(NAME(m_mask));
+	save_item(NAME(m_previous_cipherwords));
+	save_item(NAME(m_previous_plainwords));
+	save_item(NAME(m_nonlinear_count));
+
+	m_nonlinear_region_base = nonlinear_table->base();
+	m_nonlinear_region_size = nonlinear_table->bytes();
```

Comment:
Can you turn `m_nonlinear_region_base` and `m_nonlinear_region_size` into `required_memory_region m_nonlinear_region`?  Then you can get the base pointer as `m_nonlinear_region->base()` and the size as `m_nonlinear_region->bytes()` and you don’t need the glue here to get a fatal error if the region isn’t present.

Reply:
Sounds good to me, gave it test and it does everything I was going for so I changed it.